pkg/strategy/streaming.go

```diff
@@ -105,16 +100,66 @@ func newStreamingStrategy(ctx context.Context, jaeger *v1.Jaeger) S {
 		}
 	}
 
+	var indexCleaner *batchv1beta1.CronJob
 	if isBoolTrue(jaeger.Spec.Storage.EsIndexCleaner.Enabled) {
 		if strings.EqualFold(jaeger.Spec.Storage.Type, "elasticsearch") {
-			manifest.cronJobs = append(manifest.cronJobs, *cronjob.CreateEsIndexCleaner(jaeger))
+			indexCleaner = cronjob.CreateEsIndexCleaner(jaeger)
 		} else {
 			jaeger.Logger().WithField("type", jaeger.Spec.Storage.Type).Warn("Skipping Elasticsearch index cleaner job due to unsupported storage.")
 		}
 	}
 
+	var esRollover []batchv1beta1.CronJob
+	if storage.EnableRollover(jaeger.Spec.Storage) {
+		esRollover = cronjob.CreateRollover(jaeger)
+	}
+
+	// prepare the deployments, which may get changed by the elasticsearch routine
+	cDep := collector.Get()
+	queryDep := inject.Sidecar(jaeger, inject.OAuthProxy(jaeger, query.Get()))
+	var ingesterDep *appsv1.Deployment
+	if d := ingester.Get(); d != nil {
+		ingesterDep = d
+	}
 	manifest.dependencies = storage.Dependencies(jaeger)
 
+	// assembles the pieces for an elasticsearch self-provisioned deployment via the elasticsearch operator
+	if storage.ShouldDeployElasticsearch(jaeger.Spec.Storage) {
```

Comment:
Most of this new code seems to be duplicated from the production strategy. Can't we find a way to reuse it? Duplicating a couple of lines is fine, but this is a bit more than just a couple...

https://github.com/jaegertracing/jaeger-operator/blob/5999ea6f1057cc63d30888272e9f101a697da8f6/pkg/strategy/production.go#L106-L126

Reply:
I can extract this to a func, but anyways the whole streaming strategy is mostly copy of production strategy with added ingester.

Reply:
I can understand the streaming strategy code being very similar to the production when it comes to the assembled objects (deployments, services, ...), as the final set of objects is quite similar, but the logic around the self-provisioning of ES could be abstracted away. For instance, here's how the auto-provisioning of Kafka happens:

https://github.com/jaegertracing/jaeger-operator/blob/5999ea6f1057cc63d30888272e9f101a697da8f6/pkg/strategy/streaming.go#L61-L66

The strategy itself needs to know only whether the provisioning should happen. The auto-provision procedure will accept a manifest, and return a modified manifest, which might or not contain the Kafka objects required to do the provisioning.

Reply:
I have moved it to a shared func.